### PR TITLE
TLS support

### PR DESCRIPTION
**Pull request by ollls**
Submitting the code. ( branch: tls-support ). This was used for small HTTP file server example.
Please, check description:  https://github.com/ollls/zio-nio-tls-example.
Idea is that you can attach effectful TLS channel to regular NIO raw channel, all TLS operations executed as ZIO effects. Reads and writes can work with arbitrary sizes. 
Please, see: 
https://github.com/ollls/zio-nio-tls-example/blob/master/src/main/scala/example/TlsServer.scala
I provided test suit:
https://github.com/ollls/zio-nio/blob/tls-support/src/test/scala/zio/nio/channels/TlsChannelSuite.scala

Above test  creates a server and a client and copies ZIO.png file two times with extra small and large buffers respectively, then it compares files with original.

Project was tested, special attention was given to various TLS packet fragmentation during TLS handshake and TLS data transfer with different clients. Small HTTP file server based on that code can load snapshot of CNN page with any browser over TLS. General expectation is that this should be a performant solution for basic HTTP operations. Thanks to ZIO effects threads can be shared by clients to support data exchange.

### CODE REVIEW

File: src/main/scala/zio/nio/SSLEngine.scala [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import zio.blocking._
+
+import javax.net.ssl.{ SSLEngine => JSSLEngine }
+
+import javax.net.ssl.SSLEngineResult
+
+import java.nio.{ ByteBuffer => JByteBuffer }
+import java.lang.Runnable
+
+class SSLEngine(val engine: JSSLEngine) {
```
Comment (pshemass) [inline src/main/scala/zio/nio/SSLEngine.scala]: engine should be ```private``` to ```nio```

File: src/main/scala/zio/nio/channels/AsynchronousByteTlsChannel.scala [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+import zio.clock.Clock
+
+import zio.nio.SSLEngine
+
+object AsynchronousTlsByteChannel {
+
+  def apply(
+    raw_ch: AsynchronousSocketChannel,
+    sslContext: SSLContext
+  ): ZIO[Blocking with Clock, Exception, AsynchronousTlsByteChannel] = {
```
Comment (pshemass) [inline src/main/scala/zio/nio/channels/AsynchronousByteTlsChannel.scala]: use ```Managed``` like in https://github.com/zio/zio-nio/pull/68

File: src/main/scala/zio/nio/channels/AsynchronousByteTlsChannel.scala [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+      _          <- AsynchronousTlsByteChannel.open(raw_ch, ssl_engine)
+      r          <- IO.effect(new AsynchronousTlsByteChannel(raw_ch, ssl_engine))
+    } yield (r)
+
+    R.refineToOrDie[Exception]
+  }
+
+  //TLS handshake is here
+  private[nio] def open(raw_ch: AsynchronousByteChannel, ssl_engine: SSLEngine) = {
+    //println( "OPEN")
```
Comment (pshemass) [inline src/main/scala/zio/nio/channels/AsynchronousByteTlsChannel.scala]: remove commented out code

File: src/main/scala/zio/nio/channels/AsynchronousByteTlsChannel.scala [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+                _      <- IO.effect { sequential_unwrap_flag = true }
+                result <- ssl_engine.unwrap(in_buf, out_buf)
+              } yield (result.getHandshakeStatus())
+            else
+              for {
+                _ <- out_buf.clear
+
+                pos <- in_buf.position
+                lim <- in_buf.limit
+                //exhausted -  need more data.
```
Comment (pshemass) [inline src/main/scala/zio/nio/channels/AsynchronousByteTlsChannel.scala]: remove commented out code

File: src/main/scala/zio/nio/channels/AsynchronousByteTlsChannel.scala [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+
+          case NOT_HANDSHAKING => IO.succeed(NOT_HANDSHAKING)
+
+          case FINISHED => IO.succeed(FINISHED)
+
+        }
+      }
+
+      _ <- loop
+            .repeat(zio.Schedule.doWhile(c => { /*println( "do=" + c.toString );*/ c != FINISHED }))
```
Comment (pshemass) [inline src/main/scala/zio/nio/channels/AsynchronousByteTlsChannel.scala]: remove commented out code

File: src/main/scala/zio/nio/channels/AsynchronousByteTlsChannel.scala [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+    IO.effect(sslEngine.engine.getSession()).refineToOrDie[Exception]
+
+  def read(expected_size: Int): ZIO[Any with Clock, Exception, Chunk[Byte]] = {
+    val result = for {
+      in  <- Buffer.byte(expected_size)
+      out <- Buffer.byte(expected_size * 2)
+
+      nb <- channel.readBuffer(in, Duration(READ_TIMEOUT_MS, java.util.concurrent.TimeUnit.MILLISECONDS))
+
+      //p <- in.position
```
Comment (pshemass) [inline src/main/scala/zio/nio/channels/AsynchronousByteTlsChannel.scala]: remove commented out code

File: src/main/scala/zio/nio/channels/AsynchronousByteTlsChannel.scala [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+              IO.fail(new Exception("AsynchronousTlsByteChannel#read() " + res.toString()))
+            else IO.unit
+        rem <- in.remaining
+      } yield (rem)
+      _ <- loop.repeat(zio.Schedule.doWhile(_ != 0))
+      _ <- out.flip
+      //pos <- out.position
+      limit <- out.limit
+      array <- out.array
+      // _ <- IO.effect( println( "read> pos=" + pos + " " + "lim=" + limit + "sz=" + array.length ) )
```
Comment (pshemass) [inline src/main/scala/zio/nio/channels/AsynchronousByteTlsChannel.scala]: remove commented out code

File: src/main/scala/zio/nio/channels/AsynchronousByteTlsChannel.scala [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+    } yield (chunk)
+
+    result.refineToOrDie[Exception]
+  }
+
+  final def write(chunk: Chunk[Byte]): ZIO[Any with Clock, Exception, Int] = {
+    val res = for {
+
+      in <- Buffer.byte(chunk)
+      //_ <- IO.effect( println( chunk.length ) )
```
Comment (pshemass) [inline src/main/scala/zio/nio/channels/AsynchronousByteTlsChannel.scala]: remove commented out code

File: src/main/scala/zio/nio/channels/AsynchronousByteTlsChannel.scala [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+        res  <- sslEngine.wrap(in, out)
+        stat <- IO.effect(res.getStatus())
+        _ <- if (stat != SSLEngineResult.Status.OK)
+              IO.fail(new Exception("AsynchronousTlsByteChannel#write() " + res.toString()))
+            else IO.unit
+        rem <- in.remaining
+      } yield (rem)
+      _ <- loop.repeat(zio.Schedule.doWhile(_ != 0))
+      _ <- out.flip
+      //pos         <- out.position
```
Comment (pshemass) [inline src/main/scala/zio/nio/channels/AsynchronousByteTlsChannel.scala]: remove commented out code

File: src/main/scala/zio/nio/channels/AsynchronousByteTlsChannel.scala [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+  }
+
+}
+
+object AsynchronousServerTlsByteChannel {
+
+  def apply(
+    raw_ch: AsynchronousSocketChannel,
+    sslContext: SSLContext
+  ): ZIO[Blocking with Clock, Exception, AsynchronousTlsByteChannel] = {
```
Comment (pshemass) [inline src/main/scala/zio/nio/channels/AsynchronousByteTlsChannel.scala]: use ```Managed``` as in https://github.com/zio/zio-nio/pull/68

File: src/main/scala/zio/nio/channels/AsynchronousByteTlsChannel.scala [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+
+          case NOT_HANDSHAKING => IO.succeed(NOT_HANDSHAKING)
+
+          case FINISHED => IO.succeed(FINISHED)
+
+        }
+      }
+
+      r <- loop
+            .repeat(zio.Schedule.doWhile(c => { /*println( "do=" + c.toString );*/ c != FINISHED }))
```
Comment (pshemass) [inline src/main/scala/zio/nio/channels/AsynchronousByteTlsChannel.scala]: remove commented out code

File: src/test/scala/zio/nio/channels/TlsChannelSuite.scala [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+
+import zio.nio.SocketAddress
+
+import java.io.FileInputStream
+import java.io.FileOutputStream
+
+import zio.duration.Duration
+import java.util.concurrent.TimeUnit
+
+object TlsChannelSuite extends DefaultRuntime {
```
Comment (pshemass) [inline src/test/scala/zio/nio/channels/TlsChannelSuite.scala]: migrate to ```zio-tests```

**Comment by pshemass** [comment]
@ollls this looks pretty good and it will be definitely useful in ```zio-nio```. Let me know if you need help to finish this up.

File: src/main/scala/zio/nio/SSLEngine.scala (hunk 1 above)
Comment (pshemass) [reply src/main/scala/zio/nio/SSLEngine.scala]: and make it ```final```

**Comment by ollls** [comment]
Thank you! I will address those issues.

**Comment by ollls** [comment]
will resubmit on a new on a new brach, closing this one

**Pull request closed without merging**